### PR TITLE
Implement streaming encoder mode with constant memory usage.

### PR DESCRIPTION
**Pull request by szabadka**
All the internal images are now limited to the 2048x2048 DC group size, and the maximum memory usage is around 500 MB.

Benchmark results:
```
Encoding           kPixels    Bytes          BPP  E MP/s  D MP/s     Max norm  SSIMULACRA2   PSNR        pnorm       BPP*pnorm   QABPP   Bugs
---------------------------------------------------------------------------------------------------------------------------------------------
jxl:d0.1:1          156870 125499004    6.4001333   2.883  17.712   0.33339094  93.97017505  55.04   0.11910771  0.762305231867   6.400      0
jxl:d0.1:3          156870 118196370    6.0277173   1.342  14.221   0.33339094  93.97017505  55.04   0.11910771  0.717947620036   6.028      0
jxl:d0.1:4          156870 117892031    6.0121968   1.501  14.289   0.33382890  94.00462704  55.04   0.11907577  0.715906976850   6.012      0
jxl:d0.1:5          156870 139407423    7.1094276   0.777  14.252   0.34058568  93.91823707  57.02   0.10181720  0.723862019767   7.109      0
jxl:d0.1:6          156870 140721309    7.1764325   0.568  13.979   0.34023632  93.89339858  57.11   0.10187573  0.731104323064   7.176      0
jxl:d0.1:7          156870 141506387    7.2164695   0.560  14.325   0.33378362  93.90823930  57.19   0.10142266  0.731913544176   7.216      0
jxl:d0.1:1:buf3     156870 124186341    6.3331908   4.689  16.793   0.33339094  93.97017505  55.04   0.11910771  0.754331862831   6.333      0
jxl:d0.1:3:buf3     156870 118645467    6.0506202   3.936  15.506   0.33339094  93.97017505  55.04   0.11910771  0.720675522105   6.051      0
jxl:d0.1:4:buf3     156870 119051450    6.0713243   4.091  15.353   0.33382890  94.00462704  55.04   0.11907577  0.722947623654   6.071      0
jxl:d0.1:5:buf3     156870 124960819    6.3726872   1.788  13.883   0.42522492  93.86107006  55.38   0.11737111  0.747969378745   6.373      0
jxl:d0.1:6:buf3     156870 125038021    6.3766243   1.194  14.652   0.42531900  93.88796957  55.40   0.11745910  0.748992546623   6.377      0
jxl:d0.1:7:buf3     156870 125896374    6.4203982   1.156  14.181   0.43496868  93.90811739  55.47   0.11689811  0.750532425816   6.420      0
jxl:d1:1            156870 30373676    1.5489810   3.437  24.764   1.43588506  86.14940783  43.59   0.58841531  0.911444150987   2.257      0
jxl:d1:3            156870 27841570    1.4198500   2.069  26.847   1.43588506  86.14940783  43.59   0.58841531  0.835461474297   2.068      0
jxl:d1:4            156870 28110273    1.4335532   2.860  24.650   1.43953718  86.43275684  43.61   0.58675396  0.841142996644   2.092      0
jxl:d1:5            156870 23502237    1.1985549   0.913  23.353   1.49287327  84.49628039  43.63   0.58180371  0.697323707375   1.816      0
jxl:d1:6            156870 23686906    1.2079726   0.852  24.001   1.48966351  84.58298559  43.65   0.58294516  0.704181767609   1.829      0
jxl:d1:7            156870 23680833    1.2076629   0.662  24.676   1.49379655  84.59488978  43.66   0.58316869  0.704271179346   1.829      0
jxl:d1:1:buf3       156870 30144399    1.5372885   6.648  28.425   1.43588506  86.14940783  43.59   0.58841531  0.904564075601   2.240      0
jxl:d1:3:buf3       156870 28077983    1.4319065   6.575  26.224   1.43588506  86.14940783  43.59   0.58841531  0.842555684628   2.085      0
jxl:d1:4:buf3       156870 28586708    1.4578501   6.951  28.615   1.43953718  86.43275684  43.61   0.58675396  0.855399349245   2.127      0
jxl:d1:5:buf3       156870 23664692    1.2068397   2.236  25.647   1.79721683  84.31625183  43.00   0.62510227  0.754398246796   2.203      0
jxl:d1:6:buf3       156870 23700533    1.2086675   1.586  25.146   1.79445658  84.33377882  43.02   0.62494042  0.755345188024   2.201      0
jxl:d1:7:buf3       156870 23699715    1.2086258   1.441  24.907   1.79768241  84.34806803  43.03   0.62501552  0.755409881140   2.202      0
jxl:d10:1           156870  4634629    0.2363544   6.664  10.740   6.71570493  38.19963812  35.71   2.25288348  0.532478951288   1.600      0
jxl:d10:3           156870  3819655    0.1947928   5.277  12.029   6.71570493  38.19963812  35.71   2.25288348  0.438845458543   1.320      0
jxl:d10:4           156870  4137919    0.2110235   5.224  11.493   6.71726114  40.72074578  35.94   2.19382440  0.462948398734   1.427      0
jxl:d10:5           156870  3064736    0.1562938   1.756  12.541   9.14517281  30.37032353  35.23   2.69294022  0.420889917659   1.429      0
jxl:d10:6           156870  3123745    0.1593031   1.146  13.002   9.35053938  31.50052643  35.32   2.68063191  0.427033058530   1.496      0
jxl:d10:7           156870  3129105    0.1595765   0.913  13.258   9.23112087  31.60086736  35.33   2.67094507  0.426220007652   1.486      0
jxl:d10:1:buf3      156870  4788892    0.2442214   9.607  12.413   6.71570493  38.19963812  35.71   2.25288348  0.550202441230   1.652      0
jxl:d10:3:buf3      156870  4017087    0.2048613   9.304  11.483   6.71570493  38.19963812  35.71   2.25288348  0.461528694744   1.388      0
jxl:d10:4:buf3      156870  4396053    0.2241876   8.637  11.715   6.71726114  40.72074578  35.94   2.19382440  0.491828307200   1.516      0
jxl:d10:5:buf3      156870  3738546    0.1906564   2.497  11.388   7.36731798  37.57962463  35.82   2.26285872  0.431428576133   1.419      0
jxl:d10:6:buf3      156870  3790243    0.1932929   1.851  11.352   7.39668835  38.22880038  35.86   2.25680857  0.436224963822   1.442      0
jxl:d10:7:buf3      156870  3791016    0.1933323   1.679  11.463   7.40258595  38.25278127  35.86   2.25490305  0.435945531927   1.444      0
Aggregate:          156870 23301908    1.1883387   2.288  16.536   1.59684548  66.47757179  44.19   0.54200030  0.644079906108   2.696      0
```

### CODE REVIEW

File: lib/jxl/enc_ans.cc [hunk 1]
```diff
@@ -1471,6 +1496,38 @@ void ApplyLZ77(const HistogramParams& params, size_t num_contexts,
...
+                     /*log_alpha_size=*/8);
+  }
+  EncodeContextMap(context_map, codes.encoding_info.size(), writer, layer,
+                   aux_out);
+  writer->Write(1, codes.use_prefix_code);
+  size_t log_alpha_size = 8;
+  if (codes.use_prefix_code) {
+    log_alpha_size = PREFIX_MAX_BITS;
+  } else {
+    log_alpha_size = 8;  // streaming_mode
```
Comment (veluca93) [inline lib/jxl/enc_ans.cc:1515]: This comment seems to imply that this function ought to only be used in streaming mode. Is that true?

File: lib/jxl/enc_ans.cc [hunk 2]
```diff
@@ -1556,26 +1613,55 @@ size_t BuildAndEncodeHistograms(const HistogramParams& params,
...
+      if (all_singleton) {
+        use_prefix_code = true;
+      }
     }
+    codes->use_prefix_code = use_prefix_code;
+  }
+
+  if (params.add_fix_histograms) {
+    // TODO(szabadka) Add more fixed histograms.
+    const size_t alphabet_size = ANS_MAX_ALPHABET_SIZE;
```
Comment (veluca93) [inline lib/jxl/enc_ans.cc:1645]: Presumably, we can get away with smaller alphabets if we decide we are ok with, say, a 000 uint config (then 32 symbols ought to be more than enough)

File: lib/jxl/enc_ans_params.h [hunk 3]
```diff
@@ -69,6 +72,10 @@ struct HistogramParams {
   std::vector<size_t> image_widths;
   size_t max_histograms = ~0;
   bool force_huffman = false;
+  bool update_global_state = true;
+  bool streaming_mode = false;
+  bool add_missing_symbols = false;
+  bool add_fix_histograms = false;
```
Comment (veluca93) [inline lib/jxl/enc_ans_params.h:78]: s/fix/fixed/ ?

File: lib/jxl/enc_cluster.cc [hunk 4]
```diff
@@ -121,16 +147,20 @@ void FastClusterHistograms(const std::vector<Histogram>& in,
   for (size_t i = 0; i < in.size(); i++) {
     if ((*histogram_symbols)[i] != max_histograms) continue;
     size_t best = 0;
-    float best_dist = HistogramDistance(in[i], (*out)[best]);
-    for (size_t j = 1; j < out->size(); j++) {
+    float best_dist = std::numeric_limits<float>::max();
+    for (size_t j = 0; j < out->size(); j++) {
+      if (j < prev_histograms && HasNewSymbol(in[i], (*out)[j])) continue;
```
Comment (veluca93) [inline lib/jxl/enc_cluster.cc:152]: This is an optimization, but since we do not AddHistogram below, the distance in the case of j < prev_histogram should just be given by the coding cost of the new histogram with the old one - which, interestingly, should be infinity if there is a new symbol, so HasNewSymbol should not be necessary.

Otherwise, histograms such as the flat histogram are somewhat too likely to be chosen, as they have typically very few symbols compared to the actual histogram, so HistogramDistance assumes they will move much closer to the actual histogram.

File: lib/jxl/enc_frame.cc [hunk 5]
```diff
@@ -253,15 +253,16 @@ uint64_t FrameFlagsFromParams(const CompressParams& cparams) {
...
+Status LoopFilterFromParams(const CompressParams& cparams, bool streaming_mode,
                             FrameHeader* JXL_RESTRICT frame_header) {
   LoopFilter* loop_filter = &frame_header->loop_filter;
 
   // Gaborish defaults to enabled in Hare or slower.
-  loop_filter->gab = ApplyOverride(
-      cparams.gaborish, cparams.speed_tier <= SpeedTier::kHare &&
-                            frame_header->encoding == FrameEncoding::kVarDCT &&
-                            cparams.decoding_speed_tier < 4);
+  loop_filter->gab =
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:261]: TODO to remove streaming_mode from here?

File: lib/jxl/enc_frame.cc [hunk 6]
```diff
@@ -327,6 +328,8 @@ Status MakeFrameHeader(size_t xsize, size_t ysize,
   if (jpeg_data) {
     // we are transcoding a JPEG, so we don't get to choose
     frame_header->encoding = FrameEncoding::kVarDCT;
+    frame_header->x_qm_scale = 2;
+    frame_header->b_qm_scale = 2;
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:332]: I am not sure why these lines got added here...

File: lib/jxl/enc_frame.cc [hunk 7]
```diff
@@ -452,6 +456,11 @@ Status MakeFrameHeader(size_t xsize, size_t ysize,
   frame_header->animation_frame.duration = frame_info.duration;
   frame_header->animation_frame.timecode = frame_info.timecode;
 
+  if (jpeg_data) {
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:459]: or these :) (I mean, they are not incorrect, they just seem unrelated)

Comment (veluca93) [reply lib/jxl/enc_frame.cc:459]: Never mind, they seem to have been moved.

File: lib/jxl/enc_frame.cc [hunk 8]
```diff
@@ -1134,29 +1129,30 @@ Status EncodeGlobalDCInfo(const PassesSharedState& shared, BitWriter* writer,
...
-                          AuxOut* aux_out) {
+                          ModularFrameEncoder* enc_modular, AuxOut* aux_out) {
   PassesSharedState& shared = enc_state->shared;
-  JXL_RETURN_IF_ERROR(DequantMatricesEncode(
-      shared.matrices, writer, kLayerQuant, aux_out, modular_frame_encoder));
+  JXL_RETURN_IF_ERROR(DequantMatricesEncode(shared.matrices, writer,
+                                            kLayerQuant, aux_out, enc_modular));
   size_t num_histo_bits = CeilLog2Nonzero(shared.frame_dim.num_groups);
-  if (num_histo_bits != 0) {
+  if (!enc_state->streaming_mode && num_histo_bits != 0) {
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:1137]: I assume these get written somewhere else in this case? If so, a comment would be useful for us to remember in the future.

File: lib/jxl/enc_frame.cc [hunk 9]
```diff
@@ -1134,29 +1129,30 @@ Status EncodeGlobalDCInfo(const PassesSharedState& shared, BitWriter* writer,
...
-    size_t order_bits = 0;
-    JXL_RETURN_IF_ERROR(
-        U32Coder::CanEncode(kOrderEnc, enc_state->used_orders[i], &order_bits));
-    BitWriter::Allotment allotment(writer, order_bits);
-    JXL_CHECK(U32Coder::Write(kOrderEnc, enc_state->used_orders[i], writer));
-    allotment.ReclaimAndCharge(writer, kLayerOrder, aux_out);
-    EncodeCoeffOrders(enc_state->used_orders[i],
-                      &shared.coeff_orders[i * shared.coeff_order_size], writer,
-                      kLayerOrder, aux_out);
+    if (!enc_state->streaming_mode) {
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:1145]: Same here

File: lib/jxl/enc_frame.cc [hunk 10]
```diff
@@ -1167,9 +1163,27 @@ Status EncodeGlobalACInfo(PassesEncoderState* enc_state, BitWriter* writer,
     if (enc_state->cparams.decoding_speed_tier >= 1) {
       hist_params.max_histograms = 6;
     }
+    size_t num_histogram_groups = shared.num_histograms;
+    if (enc_state->streaming_mode) {
+      size_t prev_num_histograms =
+          enc_state->passes[i].codes.encoding_info.size();
+      if (enc_state->update_global_state) {
+        prev_num_histograms += kNumFixedHistograms;
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:1171]: Unless I misunderstand the code, it seems that the `update_global_state` flag here is doing "double duty" to also detect the first iteration. Is that incorrect?

File: lib/jxl/enc_frame.cc [hunk 11]
```diff
@@ -1167,9 +1163,27 @@ Status EncodeGlobalACInfo(PassesEncoderState* enc_state, BitWriter* writer,
...
+        hist_params.add_fix_histograms = true;
+      }
+      size_t remaining_histograms = kClustersLimit - prev_num_histograms;
+      size_t max_histograms = remaining_histograms < 20
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:1175]: I assume this is a heuristic and the constants could be tuned in the future :)

File: lib/jxl/enc_frame.cc [hunk 12]
```diff
@@ -1317,13 +1340,42 @@ Status EncodeGroups(const FrameHeader& frame_header,
...
+    FrameHeader& mutable_frame_header, ModularFrameEncoder& enc_modular,
     PassesEncoderState& enc_state, std::vector<BitWriter>* group_codes,
     AuxOut* aux_out) {
   JXL_ASSERT(x0 + xsize <= frame_data.xsize);
   JXL_ASSERT(y0 + ysize <= frame_data.ysize);
+  const FrameHeader& frame_header = mutable_frame_header;
+  PassesSharedState& shared = enc_state.shared;
+  shared.metadata = metadata;
+  if (enc_state.streaming_mode) {
+    shared.frame_dim.Set(xsize, ysize, /*group_size_shift=*/1,
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:1355]: I wonder if we should rename `shared.frame_dim` (later), as it seems to indicate the size of the currently-processed patch now :)

File: lib/jxl/enc_frame.cc [hunk 13]
```diff
@@ -1572,39 +1629,307 @@ bool CanDoStreamingEncoding(const CompressParams& cparams,
...
+    group_sizes->push_back(dc_group.size());
+    JXL_RETURN_IF_ERROR(AppendData(*output_processor, dc_group));
+  }
+  for (size_t i = 3; i < group_codes.size(); ++i) {
+    PaddedBytes ac_group = std::move(group_codes[i]).TakeBytes();
+    group_sizes->push_back(ac_group.size());
+    JXL_RETURN_IF_ERROR(AppendData(*output_processor, ac_group));
+  }
+  return true;
+}
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:1755]: I'll assume that pretty much all the code from `kGroupSizeOffset` to here was taken from the fast-lossless code - can we move that to some shared file? (that code would not run in standalone mode anyway)

File: lib/jxl/enc_frame.cc [hunk 14]
```diff
@@ -1572,39 +1629,307 @@ bool CanDoStreamingEncoding(const CompressParams& cparams,
...
+  size_t group_size = 256;
+  size_t dc_group_size = group_size * kBlockDim;
+  size_t dc_group_xsize = DivCeil(frame_data.xsize, dc_group_size);
+  size_t min_dc_global_size = 0;
+  size_t group_data_offset = 0;
+  PaddedBytes frame_header_bytes;
+  PaddedBytes dc_global_bytes;
+  std::vector<size_t> group_sizes;
+  size_t start_pos = output_processor->CurrentPosition();
+  for (size_t i = 0; i < dc_group_order.size(); ++i) {
```
Comment (veluca93) [inline lib/jxl/enc_frame.cc:1861]: This logic is going to be an absolute pleasure to parallelize......

**Review by veluca93 — COMMENTED** [review]
I did not find the logic that deals with DC histograms - could you point me at it?

**Comment by szabadka** [comment]
> I did not find the logic that deals with DC histograms - could you point me at it?

We set the add_missing_symbols histogram param that makes it possible for each histogram to encode any future symbols, and we call the functions that encode DC global section only once.

File: lib/jxl/dec_ans.cc [hunk 15]
```diff
@@ -354,17 +357,6 @@ Status DecodeHistograms(BitReader* br, size_t num_contexts, ANSCode* code,
   const size_t max_alphabet_size = 1 << code->log_alpha_size;
   JXL_RETURN_IF_ERROR(
       DecodeANSCodes(num_histograms, max_alphabet_size, br, code));
-  // When using LZ77, flat codes might result in valid codestreams with
-  // histograms that potentially allow very large bit counts.
-  // TODO(veluca): in principle, a valid codestream might contain a histogram
-  // that could allow very large numbers of bits that is never used during ANS
-  // decoding. There's no benefit to doing that, though.
-  if (!code->lz77.enabled && code->max_num_bits > 32) {
```
[the remark this answers is not in the document]
Comment (szabadka) [reply lib/jxl/dec_ans.cc:362]: The histograms that can encode all possible symbols were tripping it, and the comment itself said there were many false positives.

max_num_bits is used in dec_frame.cc to decide if we want a 16 bits coefficient buffer

File: lib/jxl/enc_ans.cc [hunk 16]
```diff
@@ -473,14 +473,6 @@ size_t BuildAndStoreANSEncodingData(
     return cost;
   }
   JXL_ASSERT(alphabet_size <= ANS_TAB_SIZE);
-  // Ensure we ignore trailing zeros in the histogram.
```
[the remark this answers is not in the document]
Comment (szabadka) [reply lib/jxl/enc_ans.cc:476]: alphabet_size is already largest_symbol + 1 at this point (see the newly added alphabet_size() function for Histogram)

File: lib/jxl/enc_ans.cc [hunk 17]
```diff
@@ -685,6 +677,23 @@ void ChooseUintConfigs(const HistogramParams& params,
 #endif
 }
 
+Histogram HistogramFromSymbolInfo(
```
[the remark this answers is not in the document]
Comment (szabadka) [reply lib/jxl/enc_ans.cc:680]: No, it is only used in clustering for deciding if we want to assign a context to a previous histogram.

File: lib/jxl/enc_ans.cc [hunk 18]
```diff
@@ -699,21 +708,28 @@ class HistogramBuilder {
...
+    for (size_t i = 0; i < prev_histograms; ++i) {
+      clustered_histograms.push_back(HistogramFromSymbolInfo(
+          codes->encoding_info[i], codes->use_prefix_code));
+    }
+    size_t context_offset = context_map->size();
+    context_map->resize(context_offset + histograms_.size());
     if (histograms_.size() > 1) {
       if (!ans_fuzzer_friendly_) {
         std::vector<uint32_t> histogram_symbols;
         ClusterHistograms(params, histograms_, kClustersLimit,
```
[the remark this answers is not in the document]
Comment (szabadka) [reply lib/jxl/enc_ans.cc:725]: Yes, the clustering function now preserves the existing histograms but it can assign contexts to them.

File: lib/jxl/enc_ans.cc (hunk 1 above)
Comment (szabadka) [reply lib/jxl/enc_ans.cc:1515]: Yes, I added a comment about it in the header.

File: lib/jxl/enc_ans.cc (hunk 2 above)
Comment (szabadka) [reply lib/jxl/enc_ans.cc:1647]: I added a separate TODO for this

File: lib/jxl/enc_ans_params.h (hunk 3 above)
Comment (szabadka) [reply lib/jxl/enc_ans_params.h]: Done.

File: lib/jxl/enc_cluster.cc (hunk 4 above)
Comment (szabadka) [reply lib/jxl/enc_cluster.cc:154]: Yes, good point. I want to evaluate separately so I added a TODO for now.

File: lib/jxl/enc_frame.cc (hunk 5 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:262]: Done.

File: lib/jxl/enc_frame.cc (hunk 6 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:333]: I moved these so that most of the code that sets frame header fields is in this function and then when we have to change it later based on image data those can be guarded by update_global_state flag.

File: lib/jxl/enc_frame.cc (hunk 8 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1141]: I added a comment to the function that explains this.

File: lib/jxl/enc_frame.cc (hunk 9 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1149]: See comment at function level

File: lib/jxl/enc_frame.cc (hunk 10 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1175]: Yes, maybe this is not the best name, this flag guards the things that we have to do when we first fill in structures that are then used at subsequent DC groups.

File: lib/jxl/enc_frame.cc (hunk 11 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1181]: Yes, added a comment and a TODO

File: lib/jxl/enc_frame.cc (hunk 12 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1361]: This is also used in non-streaming mode and in decoder, so choosing a name that describes all uses might be tricky (e.g. frame_or_patch_dim seems overly verbose)

File: lib/jxl/enc_frame.cc (hunk 14 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1867]: Yes, one step at a time :)

File: lib/jxl/enc_frame.cc (hunk 13 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1761]: The table lookups are used in the standalone mode to write the actual TOC, and the other functions are all slightly different. E.g. here we know the final frame header because we call this whole EncodeFrame() in the output part, while in fast lossless we don't because we call that in the input part to avoid an extra memcpy for the non-streaming input case. Also the fast lossless code has no DC groups so the padding logic is also a bit different.

File: lib/jxl/dec_ans.cc (hunk 15 above)
[the remark this answers is not in the document]
Comment (veluca93) [reply lib/jxl/dec_ans.cc:362]: Ah, I see - I believe we added the warning exactly to detect things that would accidentally fail that check, but I am ok with removing it

File: lib/jxl/enc_frame.cc (hunk 10 above)
Comment (veluca93) [reply lib/jxl/enc_frame.cc:1175]: initialize_global_state perhaps? Makes the "do it once" part clearer...

File: lib/jxl/enc_frame.cc (hunk 12 above)
Comment (veluca93) [reply lib/jxl/enc_frame.cc:1361]: I'd say `patch_dim` is fine (and in non-streaming mode, `patch` = `frame`)

File: lib/jxl/enc_frame.cc (hunk 13 above)
Comment (veluca93) [reply lib/jxl/enc_frame.cc:1761]: Probably still possible to make it a shared file, but we can do that as a follow-up

**Review by veluca93 — APPROVED** [review]

File: lib/jxl/enc_frame.cc (hunk 10 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1175]: Done.

File: lib/jxl/enc_frame.cc (hunk 12 above)
Comment (szabadka) [reply lib/jxl/enc_frame.cc:1361]: ok, but as you said it would be better in another PR

**Review by veluca93 — APPROVED** [review]

**Comment by qdwang** [comment]
Thanks for this improvement. I'd like to try this out. 

But how to enable this mode in C API? I didn't find a way to test though C API.

I tried to modify the `streaming_mode` and `add_missing_symbols` to `true`, but the memory consumption is still large.
